### PR TITLE
DEMOS-2413-basic-deliverable-migration

### PR DESCRIPTION
**Pull request by TARehman**
## Summary

Basics of deliverable migration - this is very much a work in progress and will need to be refined further, but I wanted to get the current version in for review so that I can get feedback. I'm worried that we're going to hit a wall where we won't be able to extrapolate data from PMDA eventually, but so far I'm not.

## Changes

Adds a workflow to migrate deliverables and to create submission action records for submissions that occurred, which helps the front-end to populate things correctly. Lots of the logic and open questions were added to `MIGRATION_LOGIC.md`.

## Validation

In progress. However, we have some tests in place.

- [X] Automated tests added or updated
- [X] Relevant automated tests pass
- [X] Manually tested
- [ ] Testing is not applicable

## Automated review

- [X] Run AI Code Review

### CODE REVIEW

File: data/migration/stage_pmda_for_migration/seeds/seeds.yml [hunk 1]
```diff
@@ -26,7 +16,37 @@ seeds:
...
               arguments:
                 to: source('legacy_pmda_raw', 'mdcd_demo_stus_rfrnc')
                 field: mdcd_demo_stus_cd
+
+  - name: crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id
+    columns:
+      - name: mdcd_dlvrbl_crnt_stus_cd
+        data:tests:
+          - relationships:
+              name: assert_all_mdcd_dlvrbl_crnt_stus_cd_in_crosswalk_exist_in_pmda
```
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/seeds/seeds.yml]: Good find, fixed.

File: data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_deliverable_action_submission_events.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        _legacy_id = 828
+)
+
+SELECT
+    gen_random_uuid() AS id,
+    sub_evt.creatd_dt AS action_timestamp,
+    f_deliv.id AS deliverable_id,
+    'Submitted Deliverable' AS action_type_id,
+    'Submitted' AS old_status_id,
+    'Submitted' AS new_status_id,
```
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_deliverable_action_submission_events.sql:15]: This is an excellent suggestion and it would definitely make things better. It is, however, mildly hampered by the fact that `prev_status_id` is not a real column, and is in fact, entirely made up. Dismissed.

File: data/migration/stage_pmda_for_migration/models/errors/errors_active_crosswaked_pmda_deliverables_with_no_final_demo.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+SELECT *
+FROM
+    {{ ref('deliverables_crosswalked_pmda_deliverables' ) }}
+WHERE
+    _legacy_mdcd_demo_id NOT IN (SELECT e1._legacy_mdcd_demo_id FROM {{ ref('final_demos_app_demonstration') }} AS e1)
```
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/errors/errors_active_crosswaked_pmda_deliverables_with_no_final_demo.sql]: This is a non-issue because we are selecting from the `final` table where we have already had to do multiple joins to enable functionality. However, I will add a test confirming that none of `_legacy_mdcd_demo_id` are null _just in case_.

File: data/migration/stage_pmda_for_migration/seeds/seeds.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/seeds/seeds.yml]: Dupe of previous comment

File: data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_deliverable_action_submission_events.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_deliverable_action_submission_events.sql:15]: This is an even better version of the previous suggestion, in that this one is strictly destructive in nature, rather than just hallucinatory. Dismissed.

File: data/migration/stage_pmda_for_migration/models/errors/errors_active_crosswaked_pmda_deliverables_with_no_final_demo.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/errors/errors_active_crosswaked_pmda_deliverables_with_no_final_demo.sql]: Dupe of previous comment

File: data/migration/stage_pmda_for_migration/models/cleaned/final/final_demos_app_deliverable_demonstration_type.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+SELECT
+    id AS deliverable_id,
+    demonstration_id,
+    'Migrated From PMDA' AS demonstration_type_tag_name_id
+FROM
+    {{ ref('final_demos_app_deliverable') }}
+WHERE
+    deliverable_type_id IN ('Implementation Plan', 'Monitoring Protocol')
```
Comment (connor-parke-icf) [inline data/migration/stage_pmda_for_migration/models/cleaned/final/final_demos_app_deliverable_demonstration_type.sql:8]: interesting, are these the only applicable deliverable types?

File: data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_type_cd_to_deliverable_type_id.csv [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+1,Monitoring Report
+3,Monitoring Report
+6,Monitoring Report
+12,Monitoring Report
+50,Interim Evaluation Report
+51,Evaluation Design
+52,Monitoring Report
+53,Monitoring Report
+54,Monitoring Report
+55,Monitoring Report
```
Comment (dustbuster) [inline data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_type_cd_to_deliverable_type_id.csv:12]: So good. lol

File: data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id.csv [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+7,Under CMS Review
+8,Under CMS Review
+9,Under CMS Review
+10,Accepted
+11,Upcoming
+12,Received and Filed
+13,Approved
+14,Under CMS Review
+15,Upcoming
+16,NULL
```
Comment (dustbuster) [inline data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id.csv:18]: For some reason i have a note that there might be more. but now that i look it appears 16 is the highest. 

Image

Comment (dustbuster) [reply data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id.csv:18]: And this is what i got. [idk if this helps, but i had to make my own document in confluence ](https://confluenceent.cms.gov/spaces/DEMOS/pages/1015311548/Datafixes+Cheatsheet). 

```
+----------------------------------------+ 
| mdcd_dlvrbl_stus_rfrnc                 |
+----+-----------------------------------+ 
| id | Deliverable status                |
+----+-----------------------------------+ 
| 0  | N/A                               | 
| 1  | Upcoming                          | 
| 2  | Work in Progress                  | 
| 3  | Submitted                         | 
| 4  | Requested Resubmission            | 
| 5  | Past Due                          | 
| 6  | Accepted                          | 
| 7  | Overridden                        | 
| 8  | Under Review                      | 
| 9  | In Audit                          | 
| 10 | Overridden / Accepted             | 
| 11 | Overridden / Request Resubmission | 
| 12 | Received                          | 
| 13 | Approved                          | 
| 14 | Under CMS Review                  | 
| 15 | Open-ended                        | 
| 16 | Pending Due Date Change           | 
+----+-----------------------------------+ 
```

**Comment by dustbuster** [comment]
Most of this looks pretty good. but there is a discrepancy in your Deliverable status table and what i have. 
If you feel like this is ok, then i'll not block you. But just double check if unsure.

https://github.com/Enterprise-CMCS/demos/pull/1630/changes#diff-fd5443e7b4a90309399e9b0511d6b047e040f66bd3c0dc185feea9b98e2b8287R1

**Review by dustbuster — APPROVED** [review]
See comments.

File: data/demos_data_tools/load_staged_data_to_demos_app.py [hunk 7]
```diff
@@ -221,7 +222,63 @@ class ArbitraryActionSql:
...
+            "demonstration_type_tag_name_id",
+        ],
+    ),
+    TransactionActionConfiguration("commit"),
     TriggerActionConfiguration("enable", "application", "create_phases_and_dates_for_new_application"),
+    TriggerActionConfiguration("enable", "deliverable", "trim_input_text_fields"),
+    ArbitraryActionSql(
+        "Run due date calculation",
+        f"CALL postgres_execute('{DEMOS_DDB_ATTACH_NAME}', 'CALL {APP_SCHEMA}.mark_deliverables_as_past_due()')",
+    ),
```
Comment (connor-parke-icf) [inline data/demos_data_tools/load_staged_data_to_demos_app.py:292]: nice, will be replicating this for fed comment period updates

File: data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_demonstration_finalized_demos.sql [hunk 8]
```diff
@@ -14,8 +14,8 @@ SELECT
     'CMS (OSORA)' AS clearance_level_id,
     medicaid_id,
     chip_id,
-    current_timestamp AS created_at,
-    current_timestamp AS updated_at,
+    coalesce(creatd_dt, current_timestamp) AS created_at,
```
Comment (connor-parke-icf) [inline data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_demonstration_finalized_demos.sql:18]: nice catch on including the original created/updated ats before defaulting to now

File: data/migration/stage_pmda_for_migration/models/models.yml [hunk 9]
```diff
@@ -24,6 +24,101 @@ models:
       - name: is_migrated_from_pmda
         data_type: boolean
 
+  - name: final_demos_app_deliverable_action
+    columns:
+      - name: id
+        data_type: uuid
+      - name: action_timestamp
+        data_type: timestamptz
+      - name: deliverable_id
```
Comment (connor-parke-icf) [inline data/migration/stage_pmda_for_migration/models/models.yml:45]: for these, is it necessary to validate that the ids exist in the final deliverable table?

**Review by connor-parke-icf — APPROVED** [review]
lgtm; discussed the main flow of the window functions and why theyre needed on a call. lots of good stuff to borrow for Dates and Documents later.

File: data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id.csv (hunk 6 above)
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/seeds/crosswalk_mdcd_dlvrbl_crnt_stus_cd_to_deliverable_status_id.csv:18]: So I removed 16 from this for the time being, but it is documented in `MIGRATION_LOGIC.md`. The issue is that I think that maps to something somewhat unlike what we have in our systems and I need to figure out what to do with it. It's a relatively small amount of items being filtered fortunately.

File: data/migration/stage_pmda_for_migration/models/cleaned/final/final_demos_app_deliverable_demonstration_type.sql (hunk 4 above)
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/cleaned/final/final_demos_app_deliverable_demonstration_type.sql:8]: These are the only ones that are required to have demonstration types in DEMOS so they're included here to prevent migration failures.

File: data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_demonstration_finalized_demos.sql (hunk 8 above)
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/cleaned/cleaned_demos_app_demonstration_finalized_demos.sql:18]: Yeah, I forgot to do this before, and I realized we should try and use what we can from PMDA.

File: data/migration/stage_pmda_for_migration/models/models.yml (hunk 9 above)
Comment (TARehman) [reply data/migration/stage_pmda_for_migration/models/models.yml:45]: I don't think so right now, because we're using the deliverables table to generate the action table. In _general_ yes, but in specific no.

**Review by connor-parke-icf — APPROVED** [review]
rereviewing since recent change/merge. looks good to me
